### ISSUE TITLE
CosineEmbeddingLoss

### ISSUE THREAD

**Issue by kryptonite0** [pytorch/pytorch#23782]
## 🐛 Bug

CosineEmbeddingLoss should expect a target tensor with discrete values [-1, 1], but works with any values. 

## To Reproduce
import torch
from torch.nn import CosineEmbeddingLoss

target = torch.tensor([-5.,4.])
print("target:", target)
loss = CosineEmbeddingLoss(margin=0.)
x1 = torch.tensor([[1., 0.], [0.,1.]])
x2 = torch.tensor([[0., 1.], [1.,0.]])
print("loss:", loss(x1, x2, target))

>>
target: tensor([-5.,  4.])
loss: tensor(0.)

## Expected behavior

Shouldn't I get an error message?

**Comment by tczhangzhi** [comment]
Your understanding is correct and the value of the target only makes sense when it is set to -1 or 1.
Here is the source code:
```
Tensor cosine_embedding_loss(const Tensor& input1, const Tensor& input2, const Tensor& target, double margin, int64_t reduction) {
  auto prod_sum = (input1 * input2).sum(1);
  auto mag_square1 = (input1 * input1).sum(1) + EPSILON;
  auto mag_square2 = (input2 * input2).sum(1) + EPSILON;
  auto denom = (mag_square1 * mag_square2).sqrt_();
  auto cos = prod_sum / denom;

  auto zeros = at::zeros_like(target);
  auto pos = 1 - cos;
  auto neg = (cos - margin).clamp_min_(0);
  auto output_pos = at::where(target == 1, pos, zeros);
  auto output_neg = at::where(target == -1, neg, zeros);
  auto output = output_pos + output_neg;
  return apply_loss_reduction(output, reduction);
}
```
link: https://github.com/pytorch/pytorch/blob/35fed93b1ef05175143f883c6f89f06c6dd9429b/aten/src/ATen/native/Loss.cpp#L27-L41
However, pytorch only verifies the hyperparameters without verifying the parameters. I guess the purpose of this is to improve the efficiency of the code execution. And users should ensure that the input data is correct, e.g. HingeEmbedding.
If you want to verify it yourself, you can use the following code:
```
if not (target == -1 or target == 1).all():
    # do something
    pass
```

**Comment by kryptonite0** [comment]
Thanks heaps for the explanation!
